apps/call.go

```diff
@@ -7,72 +7,106 @@ import (
-	Values        map[string]interface{} `json:"values,omitempty"`
-	Context       *Context               `json:"context,omitempty"`
-	RawCommand    string                 `json:"raw_command,omitempty"`
-	SelectedField string                 `json:"selected_field,omitempty"`
-	Query         string                 `json:"query,omitempty"`
+
+	// Type of the request, see CallType type for more information.
+	Type CallType `json:"type"`
+
+	// Values are all values entered by the user.
+	Values map[string]interface{} `json:"values,omitempty"`
+
+	// Context of execution, see the Context type for more information.
+	Context *Context `json:"context,omitempty"`
+
+	// In case the request came from the command line, the raw text of the
+	// command, as submitted by the user.
+	RawCommand string `json:"raw_command,omitempty"`
+
+	// SelectedField and Query are used in calls of type lookup, to communicate
+	// what field is selected, and what query string is already entered by the
+	// user for it.
+	SelectedField string `json:"selected_field,omitempty"`
+	Query         string `json:"query,omitempty"`
 }
 
 type CallResponseType string
 
 const (
-	// CallResponseTypeOK indicates that the call succeeded, and returns
-	// Markdown and Data.
-	CallResponseTypeOK = CallResponseType("ok")
+	// CallResponseTypeOK indicates that the call succeeded, returns optional
+	// Markdown (message) and Data.
+	CallResponseTypeOK CallResponseType = "ok"
 
-	// CallResponseTypeOK indicates an error, returns Error.
-	CallResponseTypeError = CallResponseType("error")
+	// CallResponseTypeOK indicates an error, returns ErrorText, and optional
+	// field-level errors in Data.
+	CallResponseTypeError CallResponseType = "error"
 
-	// CallResponseTypeForm returns the definition of the form to display for
-	// the inputs.
-	CallResponseTypeForm = CallResponseType("form")
+	// CallResponseTypeForm returns Form, the definition of the form to display.
+	// If returned responding to a submit, causes the form to be displayed as a
+	// modal.
+	CallResponseTypeForm CallResponseType = "form"
 
 	// CallResponseTypeCall indicates that another Call that should be executed
-	// (from the user-agent?). Call field is returned.
-	CallResponseTypeCall = CallResponseType("call")
+	// (all the way from the user-agent). Call is returned.
+	CallResponseTypeCall CallResponseType = "call"
 
 	// CallResponseTypeNavigate indicates that the user should be forcefully
 	// navigated to a URL, which may be a channel in Mattermost. NavigateToURL
 	// and UseExternalBrowser are expected to be returned.
-	// TODO should CallResponseTypeNavigate be a variation of CallResponseTypeOK?
-	CallResponseTypeNavigate = CallResponseType("navigate")
+	CallResponseTypeNavigate CallResponseType = "navigate"
 )
 
+// CallResponse is general envelope for all Call responses.
+//
+// For submit requests ok, error, form, call, or navigate response types. Form requests expect form or error. Lookup requests expect ok or error.
```

Comment:
Might be helpful to be extra verbose (helps consolidate the terminology through repetition) and instead say: "For CallRequests of type OK, error, etc.")